R/setkey.R

```diff
@@ -86,12 +86,30 @@ setkeyv = function(x, cols, verbose=getOption("datatable.verbose"), physical=TRU
     if (!typeof(.xi) %chin% c("integer","logical","character","double")) stop("Column '",i,"' is type '",typeof(.xi),"' which is not supported as a key column type, currently.")
   }
   if (!is.character(cols) || length(cols)<1L) stop("Internal error. 'cols' should be character at this point in setkey; please report.") # nocov
-  if (verbose) {
-    tt = suppressMessages(system.time(o <- forderv(x, cols, sort=TRUE, retGrp=FALSE)))  # system.time does a gc, so we don't want this always on, until refcnt is on by default in R
-    # suppress needed for tests 644 and 645 in verbose mode
-    cat("forder took", tt["user.self"]+tt["sys.self"], "sec\n")
+  
+  # get existing index name if any
+  found_index = NULL
+  if(!is.null(indices(x))) found_index <- names(attributes(attributes(x)$index))
+  new_possible_index = paste0("__", cols, collapse="")
+  
+  # forder only if index is not present
+  if(!any(new_possible_index == found_index)){
+      if (verbose) {
+          tt = suppressMessages(system.time(o <- forderv(x, cols, sort=TRUE, retGrp=FALSE)))  # system.time does a gc, so we don't want this always on, until refcnt is on by default in R
+          # suppress needed for tests 644 and 645 in verbose mode
+          cat("forder took", tt["user.self"]+tt["sys.self"], "sec\n")
+      } else {
+          o <- forderv(x, cols, sort=TRUE, retGrp=FALSE)
+      }
   } else {
-    o = forderv(x, cols, sort=TRUE, retGrp=FALSE)
+      # find the matching index
+      ix =  found_index[which(found_index == new_possible_index)]
+      if (verbose){
+          cat("using existing index for", gsub("^__","", ix), "\n")
+          o <- attr(attributes(x)$index, which=ix, exact = TRUE)
+      } else {
+          o <- attr(attributes(x)$index, which=ix, exact = TRUE)
```

Comment:
@mattdowle I've made the changes. Correct me if I am wrong, I think we need `ix` here because, let's take a scenario where a user set multiple indexes. So when we do `indices(DT)`, we get `a, b, __a__b`, if the user does `setkey(DT, a), setkey(DT, b), setkey(DT, a, b)` for a table `DT = data.table(a = c(...), b = c(...), c = c(...))` . In this case, we need to find the `ix` index of which index is matched.

Reply:
Look at line 104 though.  I commented before that line 104 could be removed.  Line 104 is, in the end, just renaming `index` as `idx`, iiuc.  This `which=` is looking up the item by name anyway.

Reply:
Done. Thanks for your patience.